mm2src/common/mm_metrics.rs

```diff
@@ -0,0 +1,814 @@
+    fn test_dashboard() {
+        let log_state = LogArc::new(LogState::in_memory());
+        let metrics = MetricsArc::new();
+
+        metrics.init_with_dashboard(log_state.weak(), 5.).unwrap();
+        let sink = metrics.sink().unwrap();
+
+        let start = sink.now();
+
+        mm_counter!(metrics, "rpc.traffic.tx", 62, "coin" => "BTC");
+        mm_counter!(metrics, "rpc.traffic.rx", 105, "coin"=> "BTC");
+
+        mm_counter!(metrics, "rpc.traffic.tx", 54, "coin" => "KMD");
+        mm_counter!(metrics, "rpc.traffic.rx", 158, "coin" => "KMD");
+
+        mm_gauge!(metrics, "rpc.connection.count", 3, "coin" => "KMD");
+
+        let end = sink.now();
+        mm_timing!(metrics,
+                   "rpc.query.spent_time",
+                   start,
+                   end,
+                   "coin" => "KMD",
+                   "method" => "blockchain.transaction.get");
+
+        block_on(async { Timer::sleep(6.).await });
+
+        mm_counter!(metrics, "rpc.traffic.tx", 30, "coin" => "BTC");
+        mm_counter!(metrics, "rpc.traffic.rx", 44, "coin" => "BTC");
+
+        mm_gauge!(metrics, "rpc.connection.count", 5, "coin" => "KMD");
+
+        let end = sink.now();
+        mm_timing!(metrics,
+                   "rpc.query.spent_time",
+                   start,
+                   end,
+                   "coin"=> "KMD",
+                   "method"=>"blockchain.transaction.get");
+
+        // measure without labels
+        mm_counter!(metrics, "test.counter", 0);
+        mm_gauge!(metrics, "test.gauge", 1);
+        let end = sink.now();
+        mm_timing!(metrics, "test.uptime", start, end);
+
+        block_on(async { Timer::sleep(6.).await });
+    }
+
+    #[test]
+    fn test_collect_json() {
+        let metrics = MetricsArc::new();
+
+        metrics.init().unwrap();
+        let mut sink = metrics.sink().unwrap();
+
+        mm_counter!(metrics, "rpc.traffic.tx", 62, "coin" => "BTC");
+        mm_counter!(metrics, "rpc.traffic.rx", 105, "coin" => "BTC");
+
+        mm_counter!(metrics, "rpc.traffic.tx", 30, "coin" => "BTC");
+        mm_counter!(metrics, "rpc.traffic.rx", 44, "coin" => "BTC");
+
+        mm_counter!(metrics, "rpc.traffic.tx", 54, "coin" => "KMD");
+        mm_counter!(metrics, "rpc.traffic.rx", 158, "coin" => "KMD");
+
+        mm_gauge!(metrics, "rpc.connection.count", 3, "coin" => "KMD");
+
+        // counter, gauge and timing may be collected also by sink API
+        sink.update_gauge_with_labels("rpc.connection.count", 5, &[("coin", "KMD")]);
+
+        mm_timing!(metrics,
+                   "rpc.query.spent_time",
+                   // ~ 1 second
+                   34381019796149, // start
+                   34382022725155, // end
+                   "coin" => "KMD",
+                   "method" => "blockchain.transaction.get");
+
+        mm_timing!(metrics,
+                   "rpc.query.spent_time",
+                   // ~ 2 second
+                   34382022774105, // start
```

Comment:
Just a question: I guess we will typically feed the mm_timing with current timestamp in nano seconds?

Reply:
Yes, the metrics crate's timing! takes nano seconds
https://docs.rs/metrics/0.12.1/metrics/macro.timing.html